source/exe/win32/platform_impl.cc

```diff
@@ -1,17 +1,56 @@
+#include <chrono>
+#include <thread>
+
+#include "common/buffer/buffer_impl.h"
 #include "common/common/assert.h"
 #include "common/common/thread_impl.h"
+#include "common/event/signal_impl.h"
 #include "common/filesystem/filesystem_impl.h"
 
 #include "exe/platform_impl.h"
 
 namespace Envoy {
 
+static volatile bool shutdown_pending = false;
+
+BOOL WINAPI CtrlHandler(DWORD fdwCtrlType) {
+  if (shutdown_pending) {
+    return 0;
+  }
+  shutdown_pending = true;
+
+  auto eventBridgeHandlers = Event::eventBridgeHandlersSingleton::get();
+  auto handler_it = eventBridgeHandlers.find(ENVOY_WIN32_SIGTERM);
+  if (handler_it == eventBridgeHandlers.end() || !handler_it->second) {
+    return 0;
+  }
+  Buffer::OwnedImpl buffer;
+  constexpr absl::string_view data{"a"};
+  buffer.add(data);
+  auto result = handler_it->second->write(buffer);
```

Comment:
A possible question: Why use a socket pair for wakeup instead of a higher level construct like a dispatcher post callback?  I think the answer is related to the set of actions that are allowed from a signal handler; writing to a socket is fine, taking locks is not and allocating memory is not fine.

Worth a comment?

Reply:
I added a comment.

 Initially my idea was that I did not want to have a global reference to the dispatcher. Seemed like a recipe for future disaster. Sadly this function does not allow for any additional context so communicating with a socket pair seems like the method that could cause the least amount of trouble. Also allocating memory or taking locks in this thread is risky.

Reply:
Right.  At least the current implementation of Dispatcher post callbacks ends up taking locks on the libevent scheduler and the dispatcher.  Also, it allocates memory.

Reply:
Also, I think you have some local changes that you have not pushed out yet.  I don't see the comments that you added.

Reply:
Fixed now in c378f5f6189624b029a4b5c17e677a25add84ce0

I was compiling the whole thing last night and did not get a chance to have a full compilation and tests so I just answered the comments, and let the code change for the next day. Sorry for the confusion